rfc/ipni-w3c.md

```diff
@@ -0,0 +1,368 @@
+
+### How web3.storage integrates IPNI today
+
+web3.storage publishes IPNI advertisements as a side-effect of the E-IPFS car [indexer-lambda].
+
+Each multihash in a CAR is sent to an SQS queue. The `publisher-lambda` takes batches from the queue, encodes and signs `Advertisement`s and writes them to a bucket as JSON.
+
+The publisher lambda knows what was the previous head to link new advertisement by reading a head file, that it previously wrote to s3 for durability. This requires configuration the sqs consumer to only allow a single instance of the lambda to run at once.
+
+The lambda makes an HTTP request to the IPNI server at `cid.contact` to inform it when the head CID of the Advertisement linked list changes.
+
+This request contains CID of the new head, w3s writes files with the CID as the filename. There is also a pre agreed `/head` file which IPNI servers can poll to see if the head has changed, and if we are still "up". If we stop responding to `/head` requests, then at some point the IPNI server may flag our content and drop it from search results and delete it from their db.
+
+The IPNI server [fetches](https://github.com/ipni/specs/blob/main/IPNI.md#advertisement-transfer) new head Advertisement from our bucket, and any others in the chain it hasn't read yet, and updates it's indexes.
+
+Our `Advertisement`s contain arbitrary batches of multihashes defined by SQS queue batching config. The `ContextID` is set to opaque bytes (a custom hash of the hashes).
+
+#### Diagram
+
+```mermaid
+flowchart TD
+    A[(dotstorage\nbucket)] -->|ObjectCreated fa:fa-car| B(bucket-to-indexer ƛ)
+    B -->|region/bucket/cid/cid.car| C[/indexer queue/]
+    C --> indexer(Indexer ƛ)
+    indexer --> |zQmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn| E[/multihash queue/]
+    E --> F(ipni Advertisement content ƛ)
+    F --> |PUT /advertCid|I
+    F --> |advert CID| G[/Advertisement queue/]
+    G --> H(ipni publish ƛ)
+    H --> |PUT /head|I[(Advert Bucket)]
+    H --> |POST head|IPNI[["`**IPNI**`"]]
+    
+    carpark[(carpark\nbucket)] --> |ObjectCreated fa:fa-car|w3infra-carpark-consumer(carpark-consumer ƛ)
+    w3infra-carpark-consumer -->|region/bucket/cid/cid.car| C[/indexer queue/]
+    indexer ---> dynamo[Dynamo\nblocks index]
+```
+
+## Proposal
+Provide a ipni/offer UCAN ability to submit the location of a CAR and a CARx-index that a user has stored with w3s, to make the them discoverable via IPFS implementations and other IPNI consumers.
+
+The ipni/offer is optional, and if not invoked, then the user's data is not indexed.
+
+If opting-in to IPNI indexing, the CAR file and the CARx-index are created by the client. The client may choose which multihashes they want indexed and specify this in the ipni/offer request.
```

Comment:
Like we do in w3filecoin protocol, we should consider here having a `ipni/accept` capability. Offer invocation asks the service to do X, service responds saying I got it and soon I will solve this for you, and by the way, here is the CID of the invocation I am working on. Then user can query for the receipt, which will be done when service ends.

A simplified version may exist when we UCAN1.0 + UCAN Invocation Spec, but I think this is the only way for it to land before that cc @Gozala

Reply:
If user wants to publish to IPNI, they invoke ipni/offer request:                                                                              
- Get blob status to check data is stored and get claims bundle CID
- Invoke `ipni/offer` with the claims bundle CID associated with the stored data to index
- User gets back a receipt that contains the advertisement ID
- Can be done once at anytime during stored data lifetime

Reply:
Once the advertisement CID is returned, then the advertisement has been published to IPNI. I don't think there is a need to make this async. The async part is waiting for IPNI to index the advertisement. That is not something we control, so the client will just have to query IPNI if they want to see when their stuff gets indexed.